### PR TITLE
ddl: refactor V2 job args of dropping column DDL.

### PR DESCRIPTION
**Pull request by joccau**
### What problem does this PR solve?

Issue Number: Ref https://github.com/pingcap/tidb/issues/53930
Ref https://github.com/pingcap/tidb/issues/54436

Problem Summary:

### What changed and how does it work?
- Refactor V2 job args for ActionDropColumn DDL.
- Set jobVersion1 at ActionDropDolumn temporarily, set version2 when refactor ActionMultiSchemaChange
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/column.go [hunk 1]
```diff
@@ -231,16 +241,37 @@ func checkDropColumn(jobCtx *jobContext, t *meta.Meta, job *model.Job) (*model.T
...
+		the caller:
+			args, err := job.DecodeArgs()
+			if err != nil {
+				return errors.trace(err)
+			}
+
+			dropColumnArgs, ok := args.(*mode.DropColumnArgs)
+			if !ok (){
+				return error.new(...)
+			}
```
Comment (D3Hunter) [inline pkg/ddl/column.go:266]: this is what XXXArgs do, why add it here

File: pkg/meta/model/job.go [hunk 2]
```diff
@@ -862,6 +874,7 @@ func (job *Job) GetInvolvingSchemaInfo() []InvolvingSchemaInfo {
 // SubJob is a representation of one DDL schema change. A Job may contain zero
 // (when multi-schema change is not applicable) or more SubJobs.
 type SubJob struct {
+	Version     JobVersion      `json:"version"`
```
Comment (D3Hunter) [inline pkg/meta/model/job.go:877]: don't add it, we only need 1 version for a job

can you change all job types of multiple schema change in this PR

File: pkg/meta/model/job_args.go [hunk 3]
```diff
@@ -467,3 +467,46 @@ func GetResourceGroupArgs(job *Job) (*ResourceGroupArgs, error) {
...
+	ColName  model.CIStr `json:"column_name,omitempty"`
+	IfExists bool        `json:"if_exists,omitempty"`
+	// indexIDs is used to make sure we don't truncate args when decoding the rawArgs.
+	IndexIDs     []int64 `json:"index_ids,omitempty"`
+	PartitionIDs []int64 `json:"partition_ids,omitempty"`
+}
+
+func (a *DropColumnArgs) fillJob(job *Job) {
+	if job.Version <= JobVersion1 {
+		job.Args = []any{a.ColName, a.IfExists, a.IndexIDs, a.PartitionIDs}
```
Comment (D3Hunter) [inline pkg/meta/model/job_args.go:482]: v1 only have 2 args

File: pkg/meta/model/job_args.go [hunk 4]
```diff
@@ -467,3 +467,46 @@ func GetResourceGroupArgs(job *Job) (*ResourceGroupArgs, error) {
 	}
 	return getOrDecodeArgsV2[*ResourceGroupArgs](job)
 }
+
+// DropColumnArgs is the arguments of dropping column job.
+type DropColumnArgs struct {
+	ColName  model.CIStr `json:"column_name,omitempty"`
+	IfExists bool        `json:"if_exists,omitempty"`
+	// indexIDs is used to make sure we don't truncate args when decoding the rawArgs.
```
Comment (D3Hunter) [inline pkg/meta/model/job_args.go:475]:
```suggestion
	// indexIDs is used to make sure we don't truncate args when decoding the rawArgs.
	// below 2 fields are filled during running.
```

File: pkg/meta/model/job.go [hunk 5]
```diff
@@ -571,8 +583,8 @@ func (job *Job) DecodeArgs(args ...any) error {
 // String implements fmt.Stringer interface.
 func (job *Job) String() string {
 	rowCount := job.GetRowCount()
-	ret := fmt.Sprintf("ID:%d, Type:%s, State:%s, SchemaState:%s, SchemaID:%d, TableID:%d, RowCount:%d, ArgLen:%d, start time: %v, Err:%v, ErrCount:%d, SnapshotVersion:%v, Version: %s",
-		job.ID, job.Type, job.State, job.SchemaState, job.SchemaID, job.TableID, rowCount, len(job.Args), TSConvert2Time(job.StartTS), job.Error, job.ErrorCount, job.SnapshotVer, job.Version)
+	ret := fmt.Sprintf("ID:%d, Type:%s, State:%s, SchemaState:%s, SchemaID:%d, TableID:%d, RowCount:%d, ArgLen:%d, RawArgsLen:%d, start time: %v, Err:%v, ErrCount:%d, SnapshotVersion:%v, Version: %s",
+		job.ID, job.Type, job.State, job.SchemaState, job.SchemaID, job.TableID, rowCount, len(job.Args), len(job.RawArgs), TSConvert2Time(job.StartTS), job.Error, job.ErrorCount, job.SnapshotVer, job.Version)
```
Comment (D3Hunter) [inline pkg/meta/model/job.go:587]: RawArgs is a byte slice, why print the length?

Comment (joccau) [reply pkg/meta/model/job.go:587]: Previously used for debugging.

Comment (joccau) [reply pkg/meta/model/job.go]: Restored

File: pkg/meta/model/job_args.go (hunk 4 above)
Comment (joccau) [reply pkg/meta/model/job_args.go:489]: Add it.

**Review by D3Hunter — APPROVED** [review]

File: pkg/meta/model/job.go [hunk 6]
```diff
@@ -490,39 +490,51 @@ func (job *Job) FillFinishedArgs(args FinishedJobArgs) {
...
+	if len(args) > 0 {
+		intest.Assert(len(args) == 1, "Job.Args should have only one element")
+		arg = args[0]
+	}
+
+	rawArgs, err := json.Marshal(arg)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return rawArgs, nil
```
Comment (lance6716) [inline pkg/meta/model/job.go:513]:
```suggestion
	return rawArgs, errors.Trace(err)
```

File: pkg/meta/model/job_args.go [hunk 7]
```diff
@@ -482,6 +482,50 @@ func GetResourceGroupArgs(job *Job) (*ResourceGroupArgs, error) {
 	return getOrDecodeArgsV2[*ResourceGroupArgs](job)
 }
 
+// DropColumnArgs is the arguments of dropping column job.
+type DropColumnArgs struct {
+	ColName  pmodel.CIStr `json:"column_name,omitempty"`
+	IfExists bool         `json:"if_exists,omitempty"`
+	// indexIDs is used to make sure we don't truncate args when decoding the rawArgs.
```
Comment (lance6716) [inline pkg/meta/model/job_args.go:489]:
```suggestion
```

Seems you copied this comment from other place, but it's just a workaround for version 1 behaviours. We don't need it now.

**Review by lance6716 — APPROVED** [review]
/hold

please unhold at your will

File: pkg/meta/model/job.go (hunk 6 above)
Comment (joccau) [reply pkg/meta/model/job.go:513]: can it be passed by lint-check? I can try it.

Comment (joccau) [reply pkg/meta/model/job.go]: modify it.

**Comment by joccau** [comment]
/test pull-br-integration-test
--

**Pull request merged**